### PR TITLE
update name has a special semantics comment

### PR DESCRIPTION
**Pull request by SergeyKanzhelev**
Related to #119

### CODE REVIEW

**Comment by bogdandrutu** [comment]
Very nicely written

**Comment by Oberon00** [comment]
So at this point, the spec begs the question: Why do we even have a method that is "highly discouraged" in the API?

**Comment by yurishkuro** [comment]
> So at this point, the spec begs the question: Why do we even have a method that is "highly discouraged" in the API?

haha, I was thinking the same yesterday. In the Jaeger clients it was pretty painful having to support late span name binding when using sampling that is sensitive to those names. I'd be curious to hear of examples where it's "completely impossible" to work around the need for late bindings. In practice, if you're able to create a span and stash is somewhere to later update the operation name, you might as well use the exact same mechanism to instead stash the parent context and a timestamp, and then create the span later when the name is known.

File: specification/api-tracing.md [hunk 1]
```diff
@@ -361,6 +361,21 @@ SHOULD be called `SetStatus`.
 Updates the `Span` name. Upon this update, any sampling behavior based on `Span`
 name will depend on the implementation.
 
+It is highly discouraged to update the name of a `Span` after it's creation.
```
Comment (c24t) [inline specification/api-tracing.md]:
```suggestion
It is highly discouraged to update the name of a `Span` after its creation.
```

File: specification/api-tracing.md [hunk 2]
```diff
@@ -361,6 +361,21 @@ SHOULD be called `SetStatus`.
...
+`Span` name is often used to group, filter and identify the logical groups of
+spans. And often, filtering logic will be implemented before the `Span` creation
+for the performance reasons. Thus the name update may interfere with this logic.
```
Comment (c24t) [inline specification/api-tracing.md]:
```suggestion
for performance reasons. Thus the name update may interfere with this logic.
```

File: specification/api-tracing.md [hunk 3]
```diff
@@ -361,6 +361,21 @@ SHOULD be called `SetStatus`.
...
+
+The method name is called `UpdateName` to differentiate this method from the
+regular property setter. Emphasizing the fact that this operation signifies a
```
Comment (c24t) [inline specification/api-tracing.md]:
```suggestion
regular property setter. It emphasizes that this operation signifies a
```

Or another change so this isn't a fragment.

File: specification/api-tracing.md [hunk 4]
```diff
@@ -361,6 +361,21 @@ SHOULD be called `SetStatus`.
...
+major change for a `Span` and may lead to re-calculation of a sampling or
+filtering decisions made previously depending on the implementation.
+
+Alternatives for the name update may be late `Span` creation, when Span is
+Started with the explicit timestamp from the past at the moment where the final
```
Comment (c24t) [inline specification/api-tracing.md]:
```suggestion
started with the explicit timestamp from the past at the moment where the final
```

File: specification/api-tracing.md [hunk 5]
```diff
@@ -361,6 +361,21 @@ SHOULD be called `SetStatus`.
...
 name will depend on the implementation.
 
+It is highly discouraged to update the name of a `Span` after it's creation.
+`Span` name is often used to group, filter and identify the logical groups of
+spans. And often, filtering logic will be implemented before the `Span` creation
+for the performance reasons. Thus the name update may interfere with this logic.
+
+The method name is called `UpdateName` to differentiate this method from the
+regular property setter. Emphasizing the fact that this operation signifies a
+major change for a `Span` and may lead to re-calculation of a sampling or
```
Comment (c24t) [inline specification/api-tracing.md]:
```suggestion
major change for a `Span` and may lead to re-calculation of sampling or
```

File: specification/api-tracing.md [hunk 6]
```diff
@@ -361,6 +361,21 @@ SHOULD be called `SetStatus`.
...
+filtering decisions made previously depending on the implementation.
+
+Alternatives for the name update may be late `Span` creation, when Span is
+Started with the explicit timestamp from the past at the moment where the final
+`Span` name is known. Or reporting a `Span` with the desired name as a child
```
Comment (c24t) [inline specification/api-tracing.md]:
```suggestion
`Span` name is known, or reporting a `Span` with the desired name as a child
```

File: specification/api-tracing.md (hunk 1 above)
Comment (c24t) [reply specification/api-tracing.md]: Or "it's created".

**Comment by Oberon00** [comment]
@yurishkuro 

> In practice, if you're able to create a span and stash is somewhere to later update the operation name, you might as well use the exact same mechanism to instead stash the parent context and a timestamp, and then create the span later when the name is known.

Actually, I just realized that there is one problem with that, namely that any new children that are created with implicit parent between taking the timestamp and starting the span get the wrong parent.

If we had an API like `finalizeStartData` that signals that the data is now ready for sampling (and maybe early partial export if an API-implementation wants to support such a thing), that might be helpful.

**Comment by SergeyKanzhelev** [comment]
@yurishkuro an example where it's hard to do is ASP.NET. You have three callbacks - Start, End and (maybe) Route identified. Between Start and Route identified other important things may happen. Also you want to represent an incoming request as a single span.

Another example is any system where the route identification is an app logic. App developer wants automatically collected requests, but for some of them she needs to set a proper route name.  

Without the `UpdateName` it will be very hard to implement either of these two scenarios. Again, with the requirement that you want to represent a single incoming request as a single `Span`.

**Comment by SergeyKanzhelev** [comment]
@c24t as usual very much appreciate your suggestions.

**Pull request merged**